Sources/Valkey/Subscriptions/SubscriptionConnectionStateMachine.swift

```diff
@@ -0,0 +1,258 @@
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import Logging
+import Synchronization
+import _ValkeyConnectionPool
+
+@available(valkeySwift 1.0, *)
+extension ValkeyClient {
+    /// Run operation with the valkey subscription connection
+    ///
+    /// - Parameters:
+    ///   - isolation: Actor isolation
+    ///   - operation: Closure to run with subscription connection
+    @usableFromInline
+    func withSubscriptionConnection<Value>(
+        isolation: isolated (any Actor)? = #isolation,
+        _ operation: (ValkeyConnection) async throws -> sending Value
+    ) async throws -> sending Value {
+        let id = self.subscriptionConnectionIDGenerator.next()
+
+        let connection = try await withTaskCancellationHandler {
+            return try await withCheckedThrowingContinuation { (cont: CheckedContinuation<ValkeyConnection, Error>) in
+                self.leaseSubscriptionConnection(id: id, request: cont)
+            }
+        } onCancel: {
+            self.cancelSubscriptionConnection(id: id)
+        }
+
+        defer {
+            self.releaseSubscriptionConnection(id: id)
+        }
+        return try await operation(connection)
+    }
+
+    func leaseSubscriptionConnection(id: Int, request: CheckedContinuation<ValkeyConnection, Error>) {
+        self.logger.trace("Get subscription connection", metadata: ["valkey_subscription_connection_id": .stringConvertible(id)])
+        self.subscriptionConnectionStateMachine.withLock { stateMachine in
+            if Task.isCancelled {
+                request.resume(throwing: CancellationError())
+            }
+            switch stateMachine.get(id: id, request: request) {
+            case .startAcquire:
+                self.queueAction(.leaseSubscriptionConnection)
+            case .completeRequest(let connection):
+                request.resume(returning: connection)
+            case .doNothing:
+                break
+            }
+        }
+    }
+
+    func acquiredSubscriptionConnection(_ connection: ValkeyConnection, releaseContinuation: CheckedContinuation<Void, Never>) {
+        self.subscriptionConnectionStateMachine.withLock { stateMachine in
+            switch stateMachine.acquired(connection, releaseRequest: releaseContinuation) {
+            case .yield(let continuations):
+                for cont in continuations {
+                    cont.resume(returning: connection)
+                }
+            }
+        }
+    }
+
+    func errorAcquiringSubscriptionConnection(_ error: Error) {
+        self.subscriptionConnectionStateMachine.withLock { stateMachine in
+            switch stateMachine.errorAcquiring(error) {
+            case .yield(let continuations):
+                for cont in continuations {
+                    cont.resume(throwing: error)
+                }
+            case .doNothing:
+                break
```

Comment:
this resumes continuations inside the lock.

Reply:
fixed